### DISCUSSION TITLE
Help tracking down duplicate requests

### DISCUSSION THREAD

**Discussion by saevarb** [urql-graphql/urql#3075] [Q&A] [answered]
Hi.

I'm currently experiencing some problems with urql creating multiple execute operations, followed by immediate teardowns, several times in a row for the same query(with identical args and `key` according to urql). I have gathered that this might be sort of normal behavior for urql, and it wouldn't really be a problem for me, except that a network request gets sent and then immediately canceled, but still manages to reach my backend. It also seems like urql is doing, say, 3 execute/teardowns in a row for the same operation, but only sending a single extra request.

I think this is probably something in my own code, but I'm having a hard time tracking it down. This particular problem is happening inside a custom hook for pagination, and I've tried optimizing it to ensure that I'm not constantly recreating the args passed to `useQuery` and such.

I have attached a screenshot below with the urql dev tools and console open with the `debugExchange` added.

![image](https://user-images.githubusercontent.com/441474/225939398-c1f10b47-60ff-4aaf-818d-783eec2eff29.png)

- Green arrow: query before changing args
- Red arrow: query after changing args
- (1) A message printed when I change my query args in response to my interaction
- (2) A message printed when the query args changed(just a `useEffect` to ensure I'm not recreating the object more than once)

The new query has `key: 18981991667` and the old query has `key: 29936900185`. As you can see in the logs, the new query is torn down multiple times in a row(and the old one a single time as expected). 

Can you folks shed some light on how I might go about avoiding this? Like I said, I don't mind things happening repeatedly in a row if I can at least avoid the spurious network request that gets canceled. I do have the `dedupExchange`, but for completeness, here are my exchanges in order:

```ts
          devtoolsExchange,
          // sync
          dedupExchange,
          // sync
          refocusExchange(),
          // sync
          cache,
          // sync
          debugExchange,
          errorExchange({
            onError(error) {
              // ..
            },
          }),
          // async
          auth(ctx),
          // async
          ssrExchange,
          // async
          fetchExchange,
```
I understand that ordering may be important in some cases -- anything off here?

Thanks for reading, and thanks for urql. :pray:

**Comment by kitten** [comment, marked as answer]
This is _somewhat_ normal behaviour. However, there's a huge caveat here.

this is typically seen in React apps and occurs because React's mount and effect timing are separate.
`useQuery` will typically get a first result on mount and then immediately cancel the operation.

However, it's unusual that a network request is sent in this case. I assume it's a React app? Can you give me the version of urql and `@urql/core` you're using please?

**Reply by saevarb** [reply]
Thanks for the swift response, but I actually found the culprit soon after posting this. :face_with_head_bandage: 

I had another hook that was supposed to be refreshing this query (via the function returned from `useQuery`) in only certain cases which was being called every time, and with `fetchPolicy: "network-only"`. Sorry for the trouble!

**Reply by kitten** [reply]
Awesome! Well, I'm just glad it's not a bug I have to hunt down! 😄
